src/Surfnet/StepupSelfService/SamlStepupProviderBundle/DependencyInjection/SurfnetStepupSelfServiceSamlStepupProviderExtension.php

```diff
@@ -0,0 +1,206 @@
+<?php
+
+/**
+ * Copyright 2014 SURFnet bv
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupSelfService\SamlStepupProviderBundle\DependencyInjection;
+
+use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+use Symfony\Component\DependencyInjection\Loader;
+
+class SurfnetStepupSelfServiceSamlStepupProviderExtension extends Extension
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
+        $config = $this->processConfiguration($configuration, $configs);
+
+        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('services.yml');
+
+        foreach ($config['providers'] as $provider => $providerConfiguration) {
+            // may seem a bit strange, but this prevents casing issue when getting/setting/creating provider
+            // service definitions etc.
+            if ($provider !== strtolower($provider)) {
+                throw new InvalidConfigurationException('The provider name must be completely lowercase');
+            }
+
+            $this->loadProviderConfiguration($provider, $providerConfiguration, $config['routes'], $container);
+        }
+    }
+
+    private function loadProviderConfiguration(
+        $provider,
+        array $configuration,
+        array $routes,
+        ContainerBuilder $container
+    ) {
+        if ($container->has('gssp.provider.' . $provider)) {
+            throw new InvalidConfigurationException(sprintf('Cannot create the same provider "%s" twice', $provider));
+        }
+
+        $this->createHostedDefinitions($provider, $configuration['hosted'], $routes, $container);
+        $this->createMetadataDefinition($provider, $configuration['hosted'], $routes, $container);
+        $this->createRemoteDefinition($provider, $configuration['remote'], $container);
+
+        $stateHandlerDefinition = new Definition('Surfnet\StepupSelfService\SamlStepupProviderBundle\Saml\StateHandler', [
+            new Reference('gssp.session'),
```

Comment:
Seems to me the `gssp.session` might convey what it is better when renamed to something like `gssp.sessionbag` or something. Now it seems like a custom sessoinhandler when in fact it is a parameterbag.